File: src/utils/percentile.js
/**
 * Percentile rank of `value` within `peers`.
 *
 * `peers` MUST be sorted ascending — the cohort files from
 * scripts/preprocess-nflverse.mjs are pre-sorted so this stays an O(log n)
 * binary search on the hot path (every factor of every player).
 *
 * Returns 0–1: the fraction of the cohort at or below `value`.
 */
export function percentileRank(value, peers) {
  if (!peers || peers.length === 0) return 0.5
  let lo = 0
  let hi = peers.length
  while (lo < hi) {
    const mid = (lo + hi) >> 1
    if (peers[mid] <= value) lo = mid + 1
    else hi = mid
  }
  return lo / peers.length
}

File: apple/Packages/FCCore/Sources/FCCore/Percentile.swift
import Foundation

/// Percentile rank against a real cohort — never a fabricated distribution.
public enum Percentile {
    /// The fraction of `sortedPeers` at or below `value`, 0…1. `peers` must be
    /// sorted ascending; the cohorts are built sorted so this stays a binary
    /// search on the hot path. An empty cohort answers 0.5, as the web app
    /// does, but callers should not ask: a percentile against nobody is not a
    /// percentile, and `PlayerGrade` drops such metrics instead.
    public static func rank(_ value: Double, in sortedPeers: [Double]) -> Double {
        guard !sortedPeers.isEmpty, value.isFinite else { return 0.5 }
        var low = 0
        var high = sortedPeers.count
        while low < high {
            let mid = (low + high) >> 1
            if sortedPeers[mid] <= value { low = mid + 1 } else { high = mid }
        }
        return Double(low) / Double(sortedPeers.count)
    }
}
